fix: sanitize response data in availability API

pedroccastro opened the pull request:
## What does this PR do?
Adds ownership validation to the availability endpoint, ensuring meeting titles and OOO reasons/notes are only returned to the data owner or system admins.

### Changes
- Added ownership check comparing requester with target user
- Created `maskBusyTimes()` to strip `title` from busy time slots
- Created `maskOutOfOfficeData()` to strip `reason`, `notes`, and `emoji` from OOO data
- System admins and own-user queries retain full data access
- Updated Swagger documentation to reflect the new behavior

### Updates since last revision
- Added integration tests for availability endpoint permissioning (addresses review feedback from @keithwillcode)
- Added unit tests for `maskBusyTimes()` and `maskOutOfOfficeData()` utility functions
- Exported handler function for testability

## How should this be tested?

### Manual testing
1. User A creates a booking with a descriptive title
2. User B queries User A's availability via API → `busy` should only contain `{ start, end }`
3. User A queries own availability → should include full `title`
4. System admin queries any user → should include full data

### Automated tests
Run the new test suite:
```bash
TZ=UTC yarn test apps/api/v1/test/lib/availability/_get.test.ts apps/api/v1/test/lib/utils/sanitizeAvailabilityResponse.test.ts
```

## Human Review Checklist
- [ ] Verify ownership check logic handles both userId and username query parameters correctly
- [ ] Confirm masking functions preserve required fields (start, end, fromUser, toUser) while removing sensitive ones
- [ ] Review that system admin bypass works as expected

## Mandatory Tasks (DO NOT REMOVE)
- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a documentation change. N/A
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

---
Link to Devin run: https://app.devin.ai/sessions/3fcdcd84eac348b0943b014571bf1d10
Requested by: unknown ()
<!-- devin-review-badge-begin -->

---

<a href="https://app.devin.ai/review/calcom/cal.com/pull/26154">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    <img src="https://static.devin.ai/assets/gh-open-in-devin-review-light.svg?v=1" alt="Open with Devin">
  </picture>
</a>
<!-- devin-review-badge-end -->

pedroccastro committed:
fix: sanitize response data in availability API
Removes event titles from busy times and personal details from
out-of-office entries when querying other users availability

keithwillcode commented on apps/api/v1/pages/api/availability/_get.ts:
```diff
@@ -202,6 +207,36 @@ async function handler(req: NextApiRequest) {
     });
+
+    // Check if the requester is the owner of the data or a system admin
+    let isOwner = isSystemWideAdmin;
```
We need integration tests around this. We already have some integration tests in API v1 that do similar checks for permissioning that you can reference.

devin-ai-integration[bot] committed:
test: add integration tests for availability endpoint permissioning

Addresses review feedback requesting integration tests around the
ownership validation and data sanitization logic.

Tests cover:
- Masking busy time titles when querying another user's availability
- Masking OOO reason/emoji/notes when querying another user
- Full data access when querying own availability (by userId or username)
- Full data access for system admins querying any user
- Edge cases (empty arrays, undefined datesOutOfOffice)

Also exports handler function for testability.

Co-authored-by: Pedro Castro <pedroccastro@users.noreply.github.com>
Co-Authored-By: unknown <>

keithwillcode committed:
Merge branch 'main' into fix/sanitize-availability-response

keithwillcode commented:
DevinAI type checks and unit tests are now failing

devin-ai-integration[bot] committed:
fix: update prismaMock import path and add type guard for handler union type

- Update import from relative path to @calcom/testing/lib/__mocks__/prismaMock
- Add SingleUserAvailability type and isSingleUserAvailability type guard
- Apply type guard to all test assertions to fix TypeScript errors
- Add explicit types to mock variables to satisfy lint rules
- Replace 'as any' with @ts-expect-error comments for prismaMock calls

Co-authored-by: Pedro Castro <pedroccastro@users.noreply.github.com>
Co-Authored-By: unknown <>
